lib/Dialect/ESI/cosim/cosim_dpi_server/DpiEntryPoints.cpp

```diff
@@ -0,0 +1,213 @@
+    // a message most of the time, this is important for performance.
+
+    if (validateSvOpenArray(data, sizeof(int8_t)) != 0) {
+      printf("ERROR: DPI-func=%s line=%d event=invalid-sv-array\n", __func__,
+             __LINE__);
+      return -2;
+    }
+
+    // Detect or verify size of buffer.
+    if (*dataSize == ~0u) {
+      *dataSize = svSizeOfArray(data);
+    } else if (*dataSize > (unsigned)svSizeOfArray(data)) {
+      printf("ERROR: DPI-func=%s line %d event=invalid-size (max %d)\n",
+             __func__, __LINE__, (unsigned)svSizeOfArray(data));
+      return -3;
+    }
+    // Verify it'll fit.
+    if (msg->size() > *dataSize) {
+      printf("ERROR: Message size too big to fit in RTL buffer\n");
+      return -5;
+    }
+
+    // Copy the message data.
+    size_t i;
+    size_t msgSize = msg->size();
+    for (i = 0; i < msgSize; ++i) {
+      auto b = msg->at(i);
+      *(char *)svGetArrElemPtr1(data, i) = b;
+    }
+    // Zero out the rest of the buffer.
+    for (; i < *dataSize; ++i) {
+      *(char *)svGetArrElemPtr1(data, i) = 0;
+    }
+    // Set the output data size.
+    *dataSize = msg->size();
+    return 0;
+  } else {
+    // No message.
+    *dataSize = 0;
+    return 0;
+  }
+}
+
+// Attempt to send data to a client.
+// - return 0 on success, negative on failure (unregistered EP).
+DPI int sv2cCosimserverEpTryPut(unsigned int endpointId,
+                                // NOLINTNEXTLINE(misc-misplaced-const)
+                                const svOpenArrayHandle data, int dataSize) {
+  if (server == nullptr)
+    return -1;
+
+  if (validateSvOpenArray(data, sizeof(int8_t)) != 0) {
+    printf("ERROR: DPI-func=%s line=%d event=invalid-sv-array\n", __func__,
+           __LINE__);
+    return -2;
+  }
+
+  // Detect or verify size.
+  if (dataSize < 0) {
+    dataSize = svSizeOfArray(data);
+  } else if (dataSize > svSizeOfArray(data)) { // not enough data
+    printf("ERROR: DPI-func=%s line %d event=invalid-size limit %d array %d\n",
+           __func__, __LINE__, dataSize, svSizeOfArray(data));
+    return -3;
+  }
+
+  Endpoint::BlobPtr blob = std::make_shared<Endpoint::Blob>(dataSize);
+  // Copy the message data into 'blob'.
+  for (long i = 0; i < dataSize; ++i) {
+    blob->at(i) = *(char *)svGetArrElemPtr1(data, i);
+  }
+  // Queue the blob.
+  Endpoint *ep = server->endpoints[endpointId];
+  if (!ep) {
+    fprintf(stderr, "Endpoint not found in registry!\n");
+    return -4;
+  }
+  ep->pushMessageToClient(blob);
+  return 0;
+}
+
+// Teardown cosimserver (disconnects from primary server port, stops connections
+// from active clients).
+DPI void sv2cCosimserverFini() {
+  printf("[cosim] Tearing down RPC server.\n");
+  if (server != nullptr) {
+    server->stop();
+    server = nullptr;
+  }
+}
+
+// Start cosimserver (spawns server for RTL-initiated work, listens for
+// connections from new SW-clients).
+DPI int sv2cCosimserverInit() {
+  if (server == nullptr) {
```

Comment:
If we assume that the DPI calls are executed in a serial context then no. But it just occurred to me that this might not be the case all the time. Good catch!